app/tasks/maintenance_tasks/task.rb

```diff
@@ -74,9 +83,57 @@ def load_constants
       end
     end
 
+    # @api private
+    ActiveRecordEnumeratorBuilder = Struct.new(:relation) do
+      def enumerator(context:)
+        JobIteration::EnumeratorBuilder.new(nil).active_record_on_records(
+          relation,
+          cursor: context.cursor,
+        )
+      end
+    end
+    private_constant :ActiveRecordEnumeratorBuilder
+
+    # @api private
+    ArrayEnumeratorBuilder = Struct.new(:array) do
+      def enumerator(context:)
+        JobIteration::EnumeratorBuilder.new(nil).build_array_enumerator(
+          array,
+          cursor: context.cursor,
+        )
+      end
+    end
+    private_constant :ArrayEnumeratorBuilder
+
+    # Returns an Enumerator builder wrapping the collection. Supported
+    # collection types are ActiveRecord::Relations, or Arrays.
+    #
+    # Tasks may override this to return a custom Enumerator builder, when
+    # requiring enumeration over an unsupported collection type. The object
+    # returned must respond to .enumerator(context:), and return an Enumerator
+    # yielding pairs of items and their cursor. The context object provided to
+    # .enumerator will respond to .cursor, returning the previous item's cursor,
+    # or nil (if this is the first iteration), to support resuming enumeration.
+    def enumerator_builder
+      collection = self.collection
+
+      case collection
+      when ActiveRecord::Relation
+        ActiveRecordEnumeratorBuilder.new(collection)
+      when Array
+        ArrayEnumeratorBuilder.new(collection)
```

Comment:
I'm not sure I like this. Previously we were calling collection and:
* creating an enumerator builder
* creating an enumerator with it, passing it the collection and cursor


Now we are:
* creating a enumerator builder with the collection
* calling enumerator on it with the context/cursor, which in turn is
* creating an enumerator builder
* creating an enumerator with the collection/cursor

I get that it's basically an adapter pattern, and it's neat for CSV to not be leaking here, but maybe it's not worth wrapping the ActiveRecord and Array enumerators here?

Reply:
I see what you mean, however foregoing the enumerator builder for ActiveRecord & Array would mean we'd have to have conditionals to treat those as special cases either by doing something like
- calling `collection` first and then falling back to `enumerator_builder` in `TaskJob`, or
- returning an enumerator directly from `enumerator_builder` and then checking the return type in `TaskJob`

At that point, I'd be tempted to suggest we go back to my original implementation in #307

```ruby
def collection
  ->(context:) { ...build the enumerator...}
end
```

I think the reason we decided not to go with that, and to instead look at the intermediate enumerator builder abstraction was so that **only authors of custom enumerators would have to know about cursors**, and not authors of tasks simply consuming such enumerators, unlike JobIteration's API, which as @rafaelfranca pointed out requires everyone to know about cursors, even for simple cases:

```ruby
def build_enumerator(cursor:)
  enumerator_builder.array([1, 2, 3], cursor: cursor)
end
```

It seems a shame to introduce an API to address those concerns, while sidestepping it internally and fragmenting the flow.

Reply:
Yeah I agree with Sam, although there's this additional cost of complexity in introducing this new abstraction, it's the price we pay to keep our simple, existing API intact while enabling custom enumeration. IMHO it's nice to have our internal enumerator builders conform to a single interface